lib/skate_web/auth_manager/error_handler.ex

```diff
@@ -7,16 +7,6 @@ defmodule SkateWeb.AuthManager.ErrorHandler do
 
   @impl Guardian.Plug.ErrorHandler
   def auth_error(conn, {_type, _reason}, _opts) do
-    auth_retries = get_session(conn, :auth_retries) || 3
-
-    if auth_retries > 0 do
-      conn
-      |> put_session(:auth_retries, auth_retries - 1)
-      |> Phoenix.Controller.redirect(to: ~p"/auth/cognito")
-    else
-      conn
-      |> delete_session(:auth_retries)
-      |> send_resp(:unauthorized, "unauthorized")
-    end
+    Phoenix.Controller.redirect(conn, to: ~p"/auth/keycloak")
```

Comment:
question(non-blocking): just wondering, if we changed the redirect based on a _global_ feature flag (i.e., a feature flag that doesn't have any users, is just on or off), could we temporaroly support keycloak AND cognito so we could roll this out and release it without requiring a deploy?

Because this has been sitting a while and we're getting a bit ansty about changes building up, maybe we could merge this and use a feature flag to release both the new keycloak login AND the new logout and other UI changes all at once (and test internally in prod without affecting users).

> I guess the advantage of having it be configurable would be that then like... Lena or Ashli could flip the switch, instead of it being a deploy? And it would be easier to roll back if something goes wrong (like inspectors write in en-masse saying that they can't login)? And we could merge this PR now without having to remember that the next prod deploy will make this abruptly go live.
> -- https://github.com/mbta/skate/pull/2358#issuecomment-1907095808

I also like the idea of being able to release this without requiring a deploy.

> The question is - configurable how? Where would this switch live? If it lives in the Skate application, how does someone access it to switch it back to the old value if there are problems, given they would need to access Skate to do so?
> -- https://github.com/mbta/skate/pull/2358#issuecomment-1908193335

Since we've confirmed that we can log-in in dev blue I'm a bit more comfortable with the idea of deploying this behind a feature flag as noted at the beginning of this.
I'm thinking that if we have cognito "enabled" (globally) then the Skate team can test that keycloak is working in prod before release by explicitly going to `<skate>/auth/keycloak`.
once we cutover to keycloak, if things are failing, we can login by explicitly going to `<skate>/auth/cognito` which would allow us access to the test group panel, where we could disable the keycloak flag, so we wouldn't need out of band ways to disable it.

I'm not sure that I've thought of every edge case, but I'd like to put this up for discussion since we're running into more difficulties with how dev would like to merge things and test iteratively and design+product would like to release it all at once.


cc: @mbta/skate-developers

Reply:
I have also been thinking about this as we've discussed the plans for logout. I suspect it may be possible to do something like this with our existing test group functionality and the override feature, though I still think it's a bit of a hack. I'm going to spend some time this morning investigating and trying it out. One thing I want to make sure we're all clear on is that, while this may help coordinate rollout, this is the sort of thing where if it breaks it will probably break for everyone, so for a rollback we still need to be prepared to do a deploy.

Reply:
I agree with pretty much all of this.

The one thing I would want to keep in mind is that we have an ambitious timeline goal that we're trying to meet, so if we do anything that might cause this to go past the end of the sprint, we should let everyone else know.

Reply:
Okay with 51832764 / b2fc5410 I've got something that seems to work up on dev-blue.